File: java/src/test/java/com/hindu/calendar/validation/FullRegressionTest.java
package com.hindu.calendar.validation;

import com.hindu.calendar.core.Masa;
import com.hindu.calendar.core.Solar;
import com.hindu.calendar.core.Tithi;
import com.hindu.calendar.ephemeris.Ephemeris;
import com.hindu.calendar.model.Location;
import com.hindu.calendar.model.MasaInfo;
import com.hindu.calendar.model.SolarCalendarType;
import com.hindu.calendar.model.SolarDate;
import com.hindu.calendar.model.TithiInfo;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Full regression: every day 1900-2050 for lunisolar + all 4 solar calendars.
 * Reads C-generated CSVs and checks Java output matches exactly.
 *
 * Lunisolar: 55,152 days × 4 checks (tithi, masa, adhika, saka) = 220,608 assertions
 * Solar: 4 calendars × ~55,152 days × 3 checks (month, day, year) = ~661,824 assertions
 * Total: ~882,432 assertions
 */
class FullRegressionTest {

    private static final Path BASE = Path.of(System.getProperty("user.dir"))
            .getParent().resolve("validation").resolve("moshier");
    private static final Location LOC = Location.NEW_DELHI;

    private static Ephemeris eph;
    private static Tithi tithi;
    private static Masa masa;
    private static Solar solar;

    @BeforeAll
    static void setUp() {
        eph = new Ephemeris();
        tithi = new Tithi(eph);
        masa = new Masa(eph, tithi);
        solar = new Solar(eph);
    }

    // ===== Lunisolar: every day 1900-2050 =====

    @Test
    void testLunisolarFullRegression() throws IOException {
        Path csv = BASE.resolve("ref_1900_2050.csv");
        assertTrue(Files.exists(csv), "Missing " + csv);

        int checked = 0;
        int failures = 0;
        StringBuilder errors = new StringBuilder();

        try (BufferedReader br = Files.newBufferedReader(csv)) {
            String header = br.readLine(); // skip header
            assertNotNull(header);

            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                int y = Integer.parseInt(parts[0]);
                int m = Integer.parseInt(parts[1]);
                int d = Integer.parseInt(parts[2]);
                int expTithi = Integer.parseInt(parts[3]);
                int expMasa = Integer.parseInt(parts[4]);
                int expAdhika = Integer.parseInt(parts[5]);
                int expSaka = Integer.parseInt(parts[6]);

                TithiInfo ti = tithi.tithiAtSunrise(y, m, d, LOC);
                MasaInfo mi = masa.masaForDate(y, m, d, LOC);

                boolean ok = true;
                if (ti.tithiNum() != expTithi) {
                    errors.append(String.format("%04d-%02d-%02d tithi: got %d, expected %d%n",
                            y, m, d, ti.tithiNum(), expTithi));
                    ok = false;
                }
                if (mi.name().number() != expMasa) {
                    errors.append(String.format("%04d-%02d-%02d masa: got %d, expected %d%n",
                            y, m, d, mi.name().number(), expMasa));
                    ok = false;
                }
                if ((mi.isAdhika() ? 1 : 0) != expAdhika) {
                    errors.append(String.format("%04d-%02d-%02d adhika: got %s, expected %d%n",
                            y, m, d, mi.isAdhika(), expAdhika));
                    ok = false;
                }
                if (mi.yearSaka() != expSaka) {
                    errors.append(String.format("%04d-%02d-%02d saka: got %d, expected %d%n",
                            y, m, d, mi.yearSaka(), expSaka));
                    ok = false;
                }

                if (!ok) failures++;
                checked++;
            }
        }

        System.out.printf("Lunisolar: %d days checked, %d failures%n", checked, failures);
        if (failures > 0) {
            fail(String.format("Lunisolar: %d/%d failures:%n%s",
                    failures, checked, errors.toString().substring(0, Math.min(errors.length(), 5000))));
        }
        assertTrue(checked > 55000, "Expected 55,000+ days, got " + checked);
    }

    // ===== Solar: every day for all 4 calendars =====

    @Test
    void testTamilFullRegression() throws IOException {
        checkSolarCalendar("tamil_months_1900_2050.csv", SolarCalendarType.TAMIL);
    }

    @Test
    void testBengaliFullRegression() throws IOException {
        checkSolarCalendar("bengali_months_1900_2050.csv", SolarCalendarType.BENGALI);
    }

    @Test
    void testOdiaFullRegression() throws IOException {
        checkSolarCalendar("odia_months_1900_2050.csv", SolarCalendarType.ODIA);
    }

    @Test
    void testMalayalamFullRegression() throws IOException {
        checkSolarCalendar("malayalam_months_1900_2050.csv", SolarCalendarType.MALAYALAM);
    }

    private void checkSolarCalendar(String csvName, SolarCalendarType type) throws IOException {
        Path csv = BASE.resolve("solar").resolve(csvName);
        assertTrue(Files.exists(csv), "Missing " + csv);

        int checked = 0;
        int failures = 0;
        StringBuilder errors = new StringBuilder();

        try (BufferedReader br = Files.newBufferedReader(csv)) {
            String header = br.readLine();
            assertNotNull(header);

            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                int solarMonth = Integer.parseInt(parts[0]);
                int solarYear = Integer.parseInt(parts[1]);
                int length = Integer.parseInt(parts[2]);
                int gregYear = Integer.parseInt(parts[3]);
                int gregMonth = Integer.parseInt(parts[4]);
                int gregDay = Integer.parseInt(parts[5]);

                // Check every day in this solar month
                for (int dayNum = 1; dayNum <= length; dayNum++) {
                    // Compute Gregorian date: start + (dayNum - 1)
                    double jd = eph.gregorianToJd(gregYear, gregMonth, gregDay) + (dayNum - 1);
                    int[] gdate = eph.jdToGregorian(jd);

                    SolarDate sd = solar.gregorianToSolar(gdate[0], gdate[1], gdate[2], LOC, type);

                    boolean ok = true;
                    if (sd.month() != solarMonth) {
                        errors.append(String.format("%s %04d-%02d-%02d month: got %d, expected %d%n",
                                type, gdate[0], gdate[1], gdate[2], sd.month(), solarMonth));
                        ok = false;
                    }
                    if (sd.day() != dayNum) {
                        errors.append(String.format("%s %04d-%02d-%02d day: got %d, expected %d%n",
                                type, gdate[0], gdate[1], gdate[2], sd.day(), dayNum));
                        ok = false;
                    }
                    if (sd.year() != solarYear) {
                        errors.append(String.format("%s %04d-%02d-%02d year: got %d, expected %d%n",
                                type, gdate[0], gdate[1], gdate[2], sd.year(), solarYear));
                        ok = false;
                    }

                    if (!ok) failures++;
                    checked++;
                }
            }
        }

        System.out.printf("%s: %d days checked, %d failures%n", type, checked, failures);
        if (failures > 0) {
            fail(String.format("%s: %d/%d failures:%n%s",
                    type, failures, checked, errors.toString().substring(0, Math.min(errors.length(), 5000))));
        }
        assertTrue(checked > 54000, type + ": Expected 54,000+ days, got " + checked);
    }
}

File: rust/tests/full_regression_test.rs
/// Full regression test: reads C-generated CSVs and verifies every day 1900-2050.
/// Lunisolar: 55,152 days × 4 checks (tithi, masa, adhika, saka)
/// Solar: 4 calendars × ~1,811 month-start entries (month, year, greg start date)

use hindu_calendar::ephemeris::Ephemeris;
use hindu_calendar::model::*;
use hindu_calendar::core::{panchang, solar};

const REF_CSV: &str = include_str!("../../validation/moshier/ref_1900_2050.csv");
const TAMIL_CSV: &str = include_str!("../../validation/moshier/solar/tamil_months_1900_2050.csv");
const BENGALI_CSV: &str = include_str!("../../validation/moshier/solar/bengali_months_1900_2050.csv");
const ODIA_CSV: &str = include_str!("../../validation/moshier/solar/odia_months_1900_2050.csv");
const MALAYALAM_CSV: &str = include_str!("../../validation/moshier/solar/malayalam_months_1900_2050.csv");

#[test]
fn test_lunisolar_55152_days() {
    let mut eph = Ephemeris::new();
    let loc = Location::NEW_DELHI;

    let mut count = 0u32;
    let mut failures = 0u32;

    for line in REF_CSV.lines().skip(1) {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() < 7 { continue; }

        let year: i32 = fields[0].parse().unwrap();
        let month: i32 = fields[1].parse().unwrap();
        let day: i32 = fields[2].parse().unwrap();
        let exp_tithi: i32 = fields[3].parse().unwrap();
        let exp_masa: i32 = fields[4].parse().unwrap();
        let exp_adhika: i32 = fields[5].parse().unwrap();
        let exp_saka: i32 = fields[6].parse().unwrap();

        let ti = hindu_calendar::core::tithi::tithi_at_sunrise(&mut eph, year, month, day, &loc);
        let hd = panchang::gregorian_to_hindu(&mut eph, year, month, day, &loc);

        let got_tithi = ti.tithi_num;
        let got_masa = hd.masa.number();
        let got_adhika = if hd.is_adhika_masa { 1 } else { 0 };
        let got_saka = hd.year_saka;

        if got_tithi != exp_tithi || got_masa != exp_masa || got_adhika != exp_adhika || got_saka != exp_saka {
            if failures < 20 {
                eprintln!(
                    "FAIL {}-{:02}-{:02}: tithi={}/{} masa={}/{} adhika={}/{} saka={}/{}",
                    year, month, day,
                    got_tithi, exp_tithi,
                    got_masa, exp_masa,
                    got_adhika, exp_adhika,
                    got_saka, exp_saka
                );
            }
            failures += 1;
        }
        count += 1;
    }

    eprintln!("Lunisolar regression: {}/{} passed ({} failures)", count - failures, count, failures);
    assert_eq!(failures, 0, "Lunisolar regression: {} failures out of {} days", failures, count);
    assert_eq!(count, 55152, "Expected 55,152 days in CSV");
}

fn test_solar_month_starts(csv: &str, cal_type: SolarCalendarType, cal_name: &str) {
    let mut eph = Ephemeris::new();
    let loc = Location::NEW_DELHI;

    let mut count = 0u32;
    let mut failures = 0u32;

    for line in csv.lines().skip(1) {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() < 7 { continue; }

        let exp_month: i32 = fields[0].parse().unwrap();
        let exp_year: i32 = fields[1].parse().unwrap();
        // fields[2] = length (not checked here)
        let greg_year: i32 = fields[3].parse().unwrap();
        let greg_month: i32 = fields[4].parse().unwrap();
        let greg_day: i32 = fields[5].parse().unwrap();
        let exp_month_name: &str = fields[6].trim();

        let sd = solar::gregorian_to_solar(&mut eph, greg_year, greg_month, greg_day, &loc, cal_type);
        let got_name = solar::solar_month_name(sd.month, cal_type);

        if sd.month != exp_month || sd.year != exp_year || got_name != exp_month_name {
            if failures < 20 {
                eprintln!(
                    "FAIL {} {}-{:02}-{:02}: month={}/{} year={}/{} name={}/{}",
                    cal_name, greg_year, greg_month, greg_day,
                    sd.month, exp_month,
                    sd.year, exp_year,
                    got_name, exp_month_name
                );
            }
            failures += 1;
        }

        // Also verify that this is day 1 of the solar month
        if sd.day != 1 {
            if failures < 20 {
                eprintln!(
                    "FAIL {} {}-{:02}-{:02}: expected solar day 1, got {}",
                    cal_name, greg_year, greg_month, greg_day, sd.day
                );
            }
            failures += 1;
        }

        count += 1;
    }

    eprintln!("{} solar regression: {}/{} passed ({} failures)", cal_name, count - failures, count, failures);
    assert_eq!(failures, 0, "{} solar regression: {} failures out of {} entries", cal_name, failures, count);
    assert!(count > 1800, "Expected ~1811 {} solar entries, got {}", cal_name, count);
}

#[test]
fn test_tamil_solar_months() {
    test_solar_month_starts(TAMIL_CSV, SolarCalendarType::Tamil, "Tamil");
}

#[test]
fn test_bengali_solar_months() {
    test_solar_month_starts(BENGALI_CSV, SolarCalendarType::Bengali, "Bengali");
}

#[test]
fn test_odia_solar_months() {
    test_solar_month_starts(ODIA_CSV, SolarCalendarType::Odia, "Odia");
}

#[test]
fn test_malayalam_solar_months() {
    test_solar_month_starts(MALAYALAM_CSV, SolarCalendarType::Malayalam, "Malayalam");
}
